cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -791,12 +917,46 @@ fn insert_common_epilogues(
     stack_size: i64,
     reg_type: ir::types::Type,
     csrs: &RegisterSet,
+    isa: &dyn TargetIsa,
+    cfa_state: Option<CFAState>,
 ) {
     while let Some(ebb) = pos.next_ebb() {
         pos.goto_last_inst(ebb);
         if let Some(inst) = pos.current_inst() {
             if pos.func.dfg[inst].opcode().is_return() {
-                insert_common_epilogue(inst, stack_size, pos, reg_type, csrs);
+                // Figure out if we need to insert end-of-function-aware frame layout information.
+                let following_inst = pos
+                    .func
+                    .layout
+                    .next_ebb(ebb)
+                    .and_then(|next_ebb| pos.func.layout.first_inst(next_ebb));
+
+                if following_inst.is_some() {
+                    if let Some(ref mut frame_layout) = pos.func.frame_layout {
```

Comment:
Could you invert these two tests, that is put the test on `if let Some()` outwards, to limit the impact of fetching the next instruction when debugging isn't enabled? It would look like:

```
let following_inst = if let Some(ref mut frame_layout) = pos.func.frame_layout {
  let inst = pos
                    .func
                    .layout
                    .next_ebb(ebb)
                    .and_then(|next_ebb| pos.func.layout.first_inst(next_ebb));
  ...
  inst
} else {
  None
};
```